jax/_src/checkify.py

```diff
@@ -725,15 +773,21 @@ class JaxRuntimeError(ValueError):
   pass
 
 @check_p.def_impl
-def check_impl(*args, err_tree):
+def check_impl(*args, err_tree, debug):
+  if debug:
+    # NOOP (check will only trigger when discharged)
+    return []
   error = tree_unflatten(err_tree, args)
   exc = error.get_exception()
   if exc:
     raise JaxRuntimeError(str(exc)) from exc
   return []
 
 @check_p.def_effectful_abstract_eval
-def check_abstract_eval(*args, err_tree):
+def check_abstract_eval(*args, err_tree, debug):
+  if debug:
```

Comment:
I'm not sure if we should mark it as not effectful. Rules should still be aware of them.

I think the right approach is to have debug effects be lowerable.

Reply:
It's the same reason we have `jax.debug.print` have an effect despite it being lowerable. Other consumers of the jaxpr might want to be aware of the effects present for their own purposes. For example, when we see "ordered effects" in a jaxpr at lowering time, we thread tokens to sequence them. The lowering needs to know about that *beforehand* so it knows to have tokens at the ready.

The analogy would be that in `jax.debug_mode` , we might run things asynchronously if they have *no* error effects. So we do want to reveal that there are `debug_check`s via the jaxpr.

Reply:
> lowerable if the experimental runtime XLA checks flag is True, and make a separate effect for debug checks, WDYT?

I don't think you need to make a separate effect, this works for now.

Reply:
> The analogy would be that in jax.debug_mode , we might run things asynchronously if they have no error effects. So we do want to reveal that there are debug_checks via the jaxpr.

But either that `debug_check` will not be run (because it will never "run") in which case it should not defeat async-ness, or its effect has been discharged into an `Error` value, in which case the output error will inherently contain that effect. My prototyped debug_mode actually checked the output `Error` to decide if it would run sync or async, because you could tell from an empty Error value that no checks are active. You could imagine that check to be on the union of the Error value "effects" and jaxpr effects?

Reply:
from offline discussion: we can use effects as a cheap static analysis, and determine whether a (debug)-check is in the code without needing to checkify!